### PR TITLE
mesa: fix cross compilation to x86_64

### PR DESCRIPTION
**Pull request by alyssais**
###### Description of changes

In particular, fix cross compilation of rusticl.  Previously, Nixpkgs has not been equipped to handle non-Cargo Rust cross compilation.  That's getting increasingly important, now that we're building Mesa with Rust, and soon the kernel, and [potentially](https://github.com/systemd/systemd/issues/26640) systemd in not too long.

This is a bigger PR than I usually like to make, but it's difficult to explain each of these changes out of context, so I've bundled them into a single PR.  I'd definitely recommend reviewing commit by commit though.  The commit messages contain the details of each change.

Spiritual successor to #159064.

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [x] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by RaitoBezarius** [comment]
Result of `nixpkgs-review pr 233628` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>8 packages failed to build:</summary>
  <ul>
    <li>mesa</li>
    <li>mesa.debug (mesa.debug.debug ,mesa.debug.dev ,mesa.debug.drivers ,mesa.debug.driversdev ,mesa.debug.opencl ,mesa.debug.osmesa ,mesa.debug.spirv2dxil)</li>
    <li>mesa.dev (mesa.dev.debug ,mesa.dev.dev ,mesa.dev.drivers ,mesa.dev.driversdev ,mesa.dev.opencl ,mesa.dev.osmesa ,mesa.dev.spirv2dxil)</li>
    <li>mesa.drivers (mesa.drivers.debug ,mesa.drivers.dev ,mesa.drivers.drivers ,mesa.drivers.driversdev ,mesa.drivers.opencl ,mesa.drivers.osmesa ,mesa.drivers.spirv2dxil)</li>
    <li>mesa.driversdev (mesa.driversdev.debug ,mesa.driversdev.dev ,mesa.driversdev.drivers ,mesa.driversdev.driversdev ,mesa.driversdev.opencl ,mesa.driversdev.osmesa ,mesa.driversdev.spirv2dxil)</li>
    <li>mesa.opencl (mesa.opencl.debug ,mesa.opencl.dev ,mesa.opencl.drivers ,mesa.opencl.driversdev ,mesa.opencl.opencl ,mesa.opencl.osmesa ,mesa.opencl.spirv2dxil)</li>
    <li>mesa.osmesa (mesa.osmesa.debug ,mesa.osmesa.dev ,mesa.osmesa.drivers ,mesa.osmesa.driversdev ,mesa.osmesa.opencl ,mesa.osmesa.osmesa ,mesa.osmesa.spirv2dxil)</li>
    <li>mesa.spirv2dxil (mesa.spirv2dxil.debug ,mesa.spirv2dxil.dev ,mesa.spirv2dxil.drivers ,mesa.spirv2dxil.driversdev ,mesa.spirv2dxil.opencl ,mesa.spirv2dxil.osmesa ,mesa.spirv2dxil.spirv2dxil)</li>
  </ul>
</details>

**Comment by RaitoBezarius** [comment]
Result of `nixpkgs-review pr 233628` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>8 packages built:</summary>
  <ul>
    <li>mesa</li>
    <li>mesa.debug (mesa.debug.debug ,mesa.debug.dev ,mesa.debug.drivers ,mesa.debug.driversdev ,mesa.debug.opencl ,mesa.debug.osmesa ,mesa.debug.spirv2dxil)</li>
    <li>mesa.dev (mesa.dev.debug ,mesa.dev.dev ,mesa.dev.drivers ,mesa.dev.driversdev ,mesa.dev.opencl ,mesa.dev.osmesa ,mesa.dev.spirv2dxil)</li>
    <li>mesa.drivers (mesa.drivers.debug ,mesa.drivers.dev ,mesa.drivers.drivers ,mesa.drivers.driversdev ,mesa.drivers.opencl ,mesa.drivers.osmesa ,mesa.drivers.spirv2dxil)</li>
    <li>mesa.driversdev (mesa.driversdev.debug ,mesa.driversdev.dev ,mesa.driversdev.drivers ,mesa.driversdev.driversdev ,mesa.driversdev.opencl ,mesa.driversdev.osmesa ,mesa.driversdev.spirv2dxil)</li>
    <li>mesa.opencl (mesa.opencl.debug ,mesa.opencl.dev ,mesa.opencl.drivers ,mesa.opencl.driversdev ,mesa.opencl.opencl ,mesa.opencl.osmesa ,mesa.opencl.spirv2dxil)</li>
    <li>mesa.osmesa (mesa.osmesa.debug ,mesa.osmesa.dev ,mesa.osmesa.drivers ,mesa.osmesa.driversdev ,mesa.osmesa.opencl ,mesa.osmesa.osmesa ,mesa.osmesa.spirv2dxil)</li>
    <li>mesa.spirv2dxil (mesa.spirv2dxil.debug ,mesa.spirv2dxil.dev ,mesa.spirv2dxil.drivers ,mesa.spirv2dxil.driversdev ,mesa.spirv2dxil.opencl ,mesa.spirv2dxil.osmesa ,mesa.spirv2dxil.spirv2dxil)</li>
  </ul>
</details>

**Comment by vcunat** [comment]
I looked just at the mesa diff and there I see no potential issue.  So `glslang` is really needed as both native and non-native?

**Comment by alyssais** [comment]
> So `glslang` is really needed as both native and non-native?

Yes.  Build dep on glslangValidator, host dep on SPIRV-Tools.

**Comment by RaitoBezarius** [comment]
Confirmed `/nix/store/h5kraiprihi263lbsl19n6ssy1bl8ynp-mesa-x86_64-unknown-linux-gnu-23.0.3` built from i686-linux to x86_64-linux (I didn't have fast aarch64-linux but I can setup a M1 if this is needed).

**Review by RaitoBezarius — APPROVED** [review]
Tested mesa, cross compilation from i686 to x86_64 for mesa and diff LGTM to me.

**Comment by Ericson2314** [comment]
I have some doubts on the first commit --- @alyssais and I are talking and I may yet be persuaded :) --- but the other 3 commits look great!

**Comment by Ericson2314** [comment]
Wait I thought I relented on the first commit and this was done months ago?

**Comment by alyssais** [comment]
> Wait I thought I relented on the first commit and this was done months ago?

Nope, I agreed that it should be unified with the elaboration params and then just took forever to find the time for another round.

File: lib/systems/default.nix
Comment (alyssais) [reply lib/systems/default.nix:172]: Fixed.  There's now "rustc.platform" (old, to be deprecated — as you pointed out on Matrix a while ago, this is a bad name), and "rust.platform" (new).

**Comment by alyssais** [comment]
Rebased. Dropped the bindgen wrapper, since Meson now supports specifying bindgen clang arguments.  rust.lib is no longer completely obsolete, due to the addition of rust.lib.envVars in the meantime, which wouldn't make sense to move to lib.

File: lib/systems/default.nix
Comment (alyssais) [reply lib/systems/default.nix]: Done.

**Comment by amjoseph-nixpkgs** [comment]
I'm definitely in favor of the first commit.  If you want to put it in a separate PR I would approve that as soon as I finish rebuilding with it.  Or I can just comment here when that finishes.

The other two commits LGTM but I don't know meson very well and I'm going to have my mesa pinned back to `22.3.7` for a few more weeks (long story).

**Comment by alyssais** [comment]
> I'm definitely in favor of the first commit. If you want to put it in a separate PR I would approve that as soon as I finish rebuilding with it.

Thanks for having a look. :)

I'm still waiting to make sure that @Ericson2314 is okay with the first commit (and AIUI he's definitely okay with the rest of them already), so I don't think splitting out the first one would speed up getting it merged at all.

**Comment by alyssais** [comment]
I've rebased to fix the merge conflict, and I've [also](https://github.com/NixOS/nixpkgs/compare/576274227287a3d630245406746bcfb3a1f07c68..c0d3cbb3592694d5062b68909dc38d3a0ffd426a):

- Simplified the logic for looking up values from `args.rust` falling back to `args.rustc` falling back to a default.
- Added an assert so that it's not possible to pass `args.rust` (the new name) and `args.rustc` (the old name) at the same time.
- Fixed `rustcTargetSpec` referring to a non-existent `rustcPlatform` that was left over from a previous iteration.

All suggested by @Ericson2314.

File: lib/systems/default.nix
Comment (amjoseph-nixpkgs) [inline lib/systems/default.nix:259]: I'd really like to not use string-pattern-matching for this...

```suggestion
          final.isNone;
```

`isUefi` is added by https://github.com/NixOS/nixpkgs/pull/231951 so once that merges we can append a `|| final.isUefi` here.

What is `switch`?

Anybody targeting `nvptx` is going to have to construct their own platform object anyways; they can just set `isNoStdTarget`.

File: pkgs/build-support/rust/lib/default.nix
Comment (amjoseph-nixpkgs) [inline pkgs/build-support/rust/lib/default.nix:23]: We can't delete these without a deprecation period.

The latest release of `crate2nix` emits code that uses them.

Could you please add `lib.warn` placeholders?  We can remove them after the branch-off.

**Review by ghost — DISMISSED** [review]
We need `lib.warn` backwards-compatibility placeholders on the `toFooBar` attributes; `crate2nix` uses them (in their current location).

Otherwise looks good.

Building:
- [ ] pkgs.tests.cross.sanity

**Comment by alyssais** [comment]
> We need `lib.warn` backwards-compatibility placeholders on the `toFooBar` attributes; `crate2nix` uses them (in their current location).

There are warnless aliases in rust/lib/default.nix already?

File: pkgs/build-support/rust/lib/default.nix
Comment (alyssais) [reply pkgs/build-support/rust/lib/default.nix:23]: They're not being deleted.  They're at the bottom of the file.

File: lib/systems/default.nix
Comment (alyssais) [reply lib/systems/default.nix:259]: I'd like that too, but I don't want to change how it works in this PR, because it's off-topic.

File: pkgs/build-support/rust/lib/default.nix
Comment (amjoseph-nixpkgs) [reply pkgs/build-support/rust/lib/default.nix:23]:
```
       error: attribute 'arch' missing

       at /git/ubik/nixpkgs/pkgs/build-support/rust/lib/default.nix:70:5:

           69|   # lib.warn "`rust.${old} platform` is deprecated. Use `platform.rust.${new}` instead."
           70|     platform.rust.${new})
             |     ^
           71| {
```

Comment (amjoseph-nixpkgs) [reply pkgs/build-support/rust/lib/default.nix:23]: > They're at the bottom of the file.

All I see at the bottom of the file is

```nix
  toTargetArch = "arch";
```

but this doesn't work (see error above; this happens for any `Cargo.nix` produced by `crate2nix`).

It might be the case that `crate2nix` is applying `toTargetArch` to things you didn't expect it to.  It should stop doing that, but we should maintain the old behavior for one cycle with a `lib.warn` in order to let that migration happen.

Comment (amjoseph-nixpkgs) [reply pkgs/build-support/rust/lib/default.nix:23]: Also, I'm totally in favor of the refactoring in d51b7f8893470d998a344acfc2eefdbdf413b2cd, but it seems a little bit weird to hide it in a PR whose title is mostly about cross-compiling a C/C++ codebase (mesa).  Could you just expose whatever extra bits you need exposed for meson, and then do the cleanup in a separate PR?  I don't want waiting for `crate2nix` to switch to the new API to block `mesa` from working...

Comment (amjoseph-nixpkgs) [reply pkgs/build-support/rust/lib/default.nix:23]: If I revert the first and last chunks changed in `pkgs/build-support/rust/lib/default.nix` everything works again.

Could we split just those two chunks out into a separate PR in order to unblock things?  Yes I know that means there will be duplicate code.  But it will maintain the existing behavior.

Comment (alyssais) [reply pkgs/build-support/rust/lib/default.nix:23]: The refactoring is necessary so that we can access this data in make-derivation.nix, which is necessary to fix Mesa cross compilation.  I've been working on fixing this problem with Mesa for a year and a half.  I don't think another few days while last-minute bugs are going to make a difference.  I definitely don't want to have duplicate code, because it might start diverging, and what if something comes up and I don't have the capacity to clean it up straight away?  It's much better to get it fixed atomically.

Comment (alyssais) [reply pkgs/build-support/rust/lib/default.nix:23]: Fixed.  It had just got out of sync with all the different revisions this has gone through, and had to be updated to use the correct new attributes.

**Review by ghost — APPROVED** [review]
Rust changes look great.

Built:

- [x] `tests.cross.sanity`
- :skull_and_crossbones: `pkgsCross.gnu64_simplekernel.mesa-23.1.9`
  - Needs https://github.com/NixOS/nixpkgs/pull/265891
  - Needs https://github.com/NixOS/nixpkgs/pull/265162
  - Probably needs https://github.com/NixOS/nixpkgs/pull/240435

**Review by Mic92 — APPROVED** [review]
Looks clean.

File: lib/systems/default.nix
Comment (Ericson2314) [inline lib/systems/default.nix:181]:
```suggestion
            /**/ if rust ? platform then rust.platform.arch
```

Comment (Ericson2314) [reply lib/systems/default.nix:181]: It is a required field of `platform`

**Review by Ericson2314 — APPROVED** [review]
Just one little thing, and then it looks good to. This is very important change let's get it in ~~23.05~~ 23.11!

(I didn't really look at all the usage site, but I think those are very low risk)

File: lib/systems/default.nix
Comment (alyssais) [reply lib/systems/default.nix:181]: Right, yeah, I see now that this was how it worked before.

Comment (alyssais) [reply lib/systems/default.nix]: Fixed.

**Comment by amjoseph-nixpkgs** [comment]
> This is very important change let's get it in 23.05!

I think that ship has sailed :laughing:

**Comment by alyssais** [comment]
The sphinx build failure on x86_64-linux seems to be intermittent / load-related.  Let's try again:

@OfBorg build mesa mesa.passthru.tests stdenv stdenv.passthru.tests

**Comment by amjoseph-nixpkgs** [comment]
> The sphinx build failure on x86_64-linux seems to be intermittent / load-related. Let's try again:

I've been getting `sphinx` test failures on my machines too.  It's annoying.  I disabled the tests for that package.  This has happened before btw.

**Comment by alyssais** [comment]
OfBorg hasn't given me the aarch64 builds for days, so I'm just going to merge.  I've tested locally on x86_64-linux and aarch64-linux, and IIRC mesa is broken on Darwin anyway.  In the unlikely event there are failures, we'll catch them in staging.

**Comment by andresilva** [comment]
Seems like this PR broke all packages that use `rustc-wasm32`, namely: `lldap`, `pagefind` and `polkadot`.

**Comment by andresilva** [comment]
I found https://github.com/NixOS/nixpkgs/pull/267876 which should fix this.

**Comment by alyssais** [comment]
Fix is https://github.com/NixOS/nixpkgs/pull/268168.

**Pull request merged**
